connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/FieldPath.java

```diff
@@ -0,0 +1,377 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+/**
+ * Represents a path to a field within a structure within a Connect key/value (e.g. Struct or
+ * Map<String, Object>).
+ * <ul>
+ * <li>It follows a dotted notation to represent nested values.</li>
+ * <li>If field names contain dots, can be escaped by wrapping field names with backticks.</li>
+ * <li>If field names contain dots at wrapping positions (beginning or end of path, before or after dots), then backticks need to be
+ * escaped by backslash.</li>
+ * </ul>
+ * Paths are calculated once and cached for further access.
+ */
+public class FieldPath {
+
+    private static final String BACKTICK = "`";
+    private static final String DOT = ".";
+    public static final char BACKTICK_CHAR = '`';
+    public static final char DOT_CHAR = '.';
+    public static final char BACKSLASH_CHAR = '\\';
+
+    private static final Map<String, FieldPath> PATHS_CACHE = new HashMap<>();
+
+    private final String[] path;
+
+    public static FieldPath ofV1(String field) {
+        return of(field, FieldSyntaxVersion.V1);
+    }
+
+    public static FieldPath ofV2(String field) {
+        return of(field, FieldSyntaxVersion.V2);
+    }
+
+    /**
+     * If version is V2, then paths are cached for further access.
+     *
+     * @param field field path expression
+     * @param version  field syntax version
+     */
+    public static FieldPath of(String field, FieldSyntaxVersion version) {
+        if (field == null || field.isEmpty() || version.equals(FieldSyntaxVersion.V1)) {
+            return new FieldPath(field, version);
+        } else {
+            if (PATHS_CACHE.containsKey(field)) {
+                return PATHS_CACHE.get(field);
+            } else {
+                final FieldPath fieldPath = new FieldPath(field, version);
+                PATHS_CACHE.put(field, fieldPath);
+                return fieldPath;
+            }
+        }
+    }
+
+    FieldPath(String path, FieldSyntaxVersion version) {
+        if (path == null || path.isEmpty()) { // empty path
+            this.path = new String[] {};
+        } else {
+            switch (version) {
+                case V1: // backward compatibility
+                    this.path = new String[] {path};
+                    break;
+                case V2:
+                    // if no dots or wrapping backticks are used, then return path with single step
+                    if (!path.contains(DOT)
+                        && !(path.startsWith(BACKTICK) && path.endsWith(
+                        BACKTICK))) {
+                        this.path = new String[] {path};
+                    } else {
+                        // prepare for tracking path steps
+                        final List<String> steps = new ArrayList<>();
+                        final StringBuilder s = new StringBuilder(
+                            path); // avoid creating new string on changes
```

Comment:
```suggestion
                        final StringBuilder s = new StringBuilder(path); // avoid creating new string on changes
```

nit: unnecessary newline